pep-0692.rst

```diff
@@ -44,64 +44,33 @@ introduce proper type annotations may be considered not worth the effort. This
 can provide.
 
-Moreover, one pattern that ``**kwargs`` are especially useful for is when a
-function should accommodate optional keyword-only arguments that don't have
-default values. A need for a pattern like that can arise when values that are
-usually used as defaults to indicate no user input, such as ``None``, can be
-passed in by a user and should result in a valid, non-default behavior.
-
-As an example, let's consider a hypothetical ``HTTPClient`` class::
-
-    class HTTPClient:
-        def __init__(
-            self,
-            *,
-            auth: AuthType,
-            timeout: TimeoutType,
-            base_url: UrlType,
-            # And some more arguments...
-        ): ...
-
-        def request(
-            self,
-            *,
-            auth,
-            timeout
-            # And some more arguments...
-        ): ...
-
-        # And a bunch of other methods...
-
-In this case, the default authorization and timeout strategies are passed in
-via the constructor. However, calls to the ``request`` method can override the
-defaults. For example, passing ``auth=None`` explicitly, could mean that no
-authorization strategy should be used at all, and the same could be true for
-``timeout``. In this situation there are no obvious default values for those
-parameters.
-
-This example is based on real world use cases taken from popular `httpx
-<httpxSentinel_>`__ and `urllib3 <urllib3Sentinel_>`__ libraries. Those
-libraries deal with this problem by introducing sentinel objects that are used
-as defaults. One issue of such a solution is that if the clients subclass a
-class that is part of a public API and want to override the ``request`` method,
-they may need to import that sentinel object, which usually should not be made
-public. Using ``**kwargs`` in this example would not have this problem and is
-arguably simpler. However, currently there is no way to properly type hint such
-a solution.
-
-In addition, ``**kwargs`` can be used to reduce the amount of code needed in
-cases when there is a top-level method that is a part of a public API and it
+Moreover, ``**kwargs`` can be used to reduce the amount of code needed in
+cases when there is a top-level function that is a part of a public API and it
 calls a bunch of helper functions, all of which expect the same keyword
 arguments. Unfortunately, if those helper functions were to use ``**kwargs``,
 there is no way to properly type hint them if the keyword arguments they expect
-are of different types.
+are of different types. However, even if the keyword arguments are of the same
+type, there is no way to check if the function is being called with keyword
+names that it actually expects. All of this makes certain issues difficult to
+spot. For example, a helper function could be updated to take account of a new
+keyword argument, but the public function which uses the helper may not have it
+in its signature.
```

Comment:
I've got a feeling that there's an even shorter paragraph trying to come forward here. The key issue is that if we have e.g. `xmlopen(file, mode)` calling the builtin `open(file, mode)`, and `open()` grows a new feature (say, `bufsize=...`), we might want to be warned that `xmlopen()` also needs to be updated to add that feature. Unfortunately, if the solution us that `open()` should be declared using `(**kwds: **SomeTypeDict)`, I'm not convinced that this is altogether an improvement -- inside `open()` it's harder to access the arguments (`kwargs["filename"]` instead of just `filename`) and for the human reader the true signature is farther removed from the function definition.

Reply:
That makes sense, I agree that in the example that you're mentioning here there is no advantage to using `**kwargs` (btw, feels like it'd be quite hard to come up with an elegant solution to that). The situation I had in mind was that for example a helper which already uses `**kwargs` in its signature gets updated to deal with a new keyword (it uses `kwagrs["new_keyword"]` in its body), but a function that calls the helper doesn't take that into account - I guess the earlier paragraph would fit this example better. But anyway I've removed it altogether as it seems to be a well known use case for which static typing is helpful and it seems to me now that there is no point to highlight this one particular advantage.